Rewrite tag-related activities and fragments in Kotlin

danilo04 opened the pull request:
Fixes #1373

### Fix
Rewrite tag-related activities and fragments in Kotlin. This includes `TagsActivity`, `AddTagActivity` and `TagDialogFragment`. 

### Test

**Search**
1. Go to left panel menu
2. Tap on `Edit Tags`
3. Search for tags and verify the tags shown are the right results
4. Search for tags that don't exists, the text `No Tags Found` should be shown

**Add Tag**
1. Tap in the plus button (bottom right corner)  to add new tags
2. Add a new tag that is not in the list of tags and save
3. Tap in the plus button (bottom right corner) and try to add a tag that already exists
4. An error message with tag that already exists should be shown
5. Add an space 
6. An error message with tags cannot contain spaces should be shown

**Edit Tag**
1. Tap in one of the tags in the list of tags
2. Add a new tag that is not in the list of tags and save
3. Tap into another tag and try to add a tag that already exists
4. A Tag conflict message shown be shown
5. Tag on merge to merge the tag

**Delete Tag**
1. Tap on the trash icon of one of the tags in the list
2. If the tag does not have a note associated, the tag will be deleted immediately
3. If the tag has at least one note associated, a dialog asking if the user is sure to delete the tag should be shown
4. Tap `Yes` and the tag should be deleted

### Review
Only one developer is required to review these changes, but anyone can perform the review.

### Release
These changes do not require release notes.

danilo04 committed:
Add dependencies ktx to work with jetpack in activities and fragments

To be able to use certain constructs, such as `by viewModels()`  in
Kotlin activities and fragments, we added dependencies needed.

danilo04 committed:
Change AddTagActivity to Kotlin

Initial version of AddTagActivity as Kotlin class. The java class was
removed.

danilo04 committed:
Use ktx extension doAfterTextChanged instead of the TextWatcher listener

danilo04 committed:
Cleanup imports

danilo04 committed:
Change setup views methods to be extensions of ActivityTagAddBinding

Reduces the methods' line lengths since it does not have to include
`binding.`.

danilo04 committed:
Cleanup of findByView calls

danilo04 committed:
Remove activity parameter since it can be accessed directly

danilo04 committed:
Rename .java to .kt

danilo04 committed:
Convert TagDialogFragment from Java to Kotlin

danilo04 committed:
Make hte tag parameter in TagDialogFragment's constructor non nullable

danilo04 committed:
Create instance of TagDialogViewModel using viewModels()

In Kotlin, we can use KTX extensions to create instances of view models

danilo04 committed:
Convert if condition to when in Kotlin to handle events

danilo04 committed:
Use KTX extensions to handle text changes on edit text for tag name

danilo04 committed:
Clean imports in TagDialogFragment

danilo04 committed:
Avoid using !! operator to dereference dialog edit tag

danilo04 committed:
Use lambda directly for button listeners instead of class fields

danilo04 committed:
Remove unused fields for button listeners

danilo04 committed:
Add view binding to TagDialogFragment

danilo04 committed:
Refactor to remove state fields in the TagDialogFragment class

Reduce the number of fields that represent state in TagDialogFragment.
We keep just the binding and dialog as state fields and from them, we
can update the whole UI.

danilo04 committed:
Create a method to couple logic related to start the UI state

danilo04 committed:
Added comments to make clear how the dialog's UI is updated

danilo04 committed:
Convert TagsActivity from Java to Kotlin

danilo04 committed:
Rename .java to .kt

danilo04 committed:
Remove lint warnings in TagsActivity

danilo04 committed:
Reformat when statement to handle events in TagsActivity

danilo04 committed:
Create instance of TagsViewModel using viewModels()

danilo04 committed:
Add extensions functions to get colors in string from color resources

danilo04 committed:
Avoid using getString with a color resource and use extension function

danilo04 committed:
Define simple toast function as extension function of Activity

danilo04 committed:
Clean unused method

danilo04 committed:
Use view binding in TagsActivity

Instead of using private fields for each view component that is needed
in the activity, we use just the binding object.

danilo04 committed:
Refactor menu creation to avoid using activity fields

danilo04 committed:
Add tests for empty tags and not tags found during search

Discovered a bug during refactoring that the no tags view was showing
even if there were tags shown. The tests added check for empty list of
tags after loading and no tags found during search.

danilo04 committed:
Fix bug that was showing no tags view even though there were tags

During the refactoring, we forgot to the set the empty view of the
recycleview. This cause the empty view that shows the text "No Tags" to
be shown. We found that we didn't have tests for these cases, so we
added them in a previous commit.

danilo04 committed:
Suppress deprecated warnings for startActivityForResult for now

danilo04 committed:
Improve format in TagsActivity

ParaskP7 commented on Simplenote/src/androidTest/java/com/automattic/simplenote/integration/TagsActivityTest.kt:
```diff
@@ -30,6 +31,15 @@ import org.junit.runner.RunWith
 class TagsActivityTest : BaseUITest() {
```
Praise (❤️): I love the 2 new empty tags related tests that you added based on the bug that you found, kudos!

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/utils/BaseExt.kt:
```diff
@@ -0,0 +1,21 @@
+fun AppCompatActivity.getColorStr(@ColorRes color: Int): String {
+    val hintColor = ContextCompat.getColor(this, color)
+    val hintHexColor = hintColor.toHexString().replace("ff", "")
+    return "#$hintHexColor"
```
Question (❓): Is this `#` character a new addition, it didn't exist before, right?

danilo04 replied:
Yes, I added it because if you get a color using `getString(...)`, it will include `#`, but since we are getting the color using `getColor(...)`, it does not include it.

ParaskP7 replied:
Ah, nice, thanks for the clarification Danilo! 🙏

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/utils/BaseExt.kt:
```diff
@@ -0,0 +1,21 @@
+
+fun Int.toHexString(): String {
+    return Integer.toHexString(this)
+}
```
Minor (🔍): Consider adding a new line at the end of this file. Actually, GitHub also warns you about it, nice! 😄 

PS: Same apply to all other files. Actually, consider adding that setting to your AS so that it is done for you automatically from now on.

danilo04 replied:
🤯 , did not know about this. Thanks! Fixed in 55059c868d009e4e830bd484b0b2ae4d2e091c7a.

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/AddTagActivity.kt:
```diff
@@ -0,0 +1,130 @@
+
+        binding.setObservers()
+        binding.setupLayout()
+        binding.setupViews()
```
Praise (❤️): I love what you did with the binding extension functions and using this pattern to build the UI, we are doing the same on `WordPress` so it is great to see it happening on `Simplenote` as well, kudos!

PS: Same applies to all the other places you did the same!

danilo04 replied:
Yes, I saw that in some files we discussed before and I thought that is a great way to reduce boilerplate and reduce the line length.

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/AddTagActivity.kt:
```diff
@@ -0,0 +1,130 @@
+
+    private fun ActivityTagAddBinding.setupViews() {
+        title.text = getString(R.string.add_tag)
+        tagInput.doAfterTextChanged { s -> viewModel.updateUiState(s.toString()) }
```
Praise (❤️): I love what you did with this KTX extension function and for managing to remove the `TextWatcher`. The whole class is so much cleaner now, kudos!

PS: Same applies to all the other places you did the same!

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/TagDialogFragment.kt:
```diff
@@ -0,0 +1,151 @@
+        dialogEditTag.setTitle(R.string.dialog_tag_conflict_title)
+
+        val positiveButton = dialogEditTag.getButton(DialogInterface.BUTTON_POSITIVE)
+        positiveButton.text = getString(R.string.dialog_tag_conflict_button_positive)
```
Question (❓): What if we were explicit with the `positiveButton` visibility as well, just like we do for the `negativeButton` and `neutralButton`?

PS: Same apply for the `showDialogRenameTag()` function below.

danilo04 replied:
For the moment is not needed since the positive button is always visible, but to avoid any bugs in future changes, I agree that making it explicit would make this code less error-prone. Fixed in 045bddff1a9612962d432649dced6d1fddc31c4e.

ParaskP7 replied:
You are a rockstar, thank you! 🙏

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/TagDialogFragment.kt:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    /**
+     * Change the dialog layout to show a conflict message in which the user can merge two tags or cancel the merge
```
Minor (🔍): Consider ending the KDoc sentence with a dot (`.`).

PS: Same apply for the showDialogRenameTag() function below.

danilo04 replied:
Good catch! Fixed in b41944abc5f2a0b8a0ac1c4fca8e9c567e8010d9.

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/TagsActivity.kt:
```diff
@@ -0,0 +1,246 @@
+    public override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        _binding = ActivityTagsBinding.inflate(layoutInflater)
```
Warning (⚠️): For activities you don't need to do the binding backing field trick so that to then make the binding null during `onDestroy()`. This is only required on fragment and in case you need that field since you need access to it from outside the `onCreate(...)`, just like you did for `TagDialogFragment`.

Actually, for `AddTagActivity` you did exactly what I am suggesting, so I am not sure why you didn't do the same for this activity as well.

As such, this can become `val binding: ActivityTagsBinding = ActivityTagsBinding.inflate(layoutInflater)` and the binding fields can be removed. Doing so will propagate further refactoring down the line, mainly making some more functions an extension function of the binding.

danilo04 replied:
Good point! Fixed in d3c92932c209f99075c7bc3ead2cbfcec4a0bcf6.

ParaskP7 replied:
Amazing, many thanks Danilo! 🕺 💃

ParaskP7 reviewed (changes requested):
👋 @danilo04 !

I have reviewed and tested this PR as per the instructions, everything works as expected (ish), amazing job! 🌟 🌟 🌟 

The only problem I found during testing was during the `Edit Tag` testing, where a tag, that is not associated with any note, wouldn't update itself. Or two conflicting tags, that are not associated with any note, wouldn't merge as well. Can you double check that? Maybe this is a known bug that is already existing out in the wild as well. 🤷 

I have left a warning (⚠️), a couple of questions (❓), some minor (🔍) comments for you to consider. I am **NOT** going to approve this PR atm, since I would like for us to figure out what is wrong with the `Edit Tag` testing and maybe apply a fix on the my warning comment.

In addition, I would like to offer a couple suggestions (💡) for you to consider for when converting a Java file to Kotlin:
- Preparing the Java file, prior hitting the `Convert to Kotlin` button, might help you create a much better Kotlin file. For example, adding `@Nullable` or `NonNull` annotation wherever possible can end up helping you enormously. Try to find out what kind of refactoring or preparation you can do before hitting the `Convert to Kotlin` button.
- When converting to `ViewBinding` and removing all fields at the same time can end up being more complicated than expected. When doing that, do a commit-by-commit field conversion, with the first commit being the binding setup. For example, if that was done on `TagsActivity` you might have saved yourself the trouble of forgetting the `list.setEmptyView(empty.root)` since it would have been more obvious that this is missing.

PS: Last but not least, consider renaming the `AddActivityViewModelTest` test class to `AddTagViewModelTest`. The file name and class name differ and as such the main class with the test class differ, which makes it harder for the AS to associate the main to the test class.

danilo04 committed:
End KDoc comments with a dot (.)

danilo04 committed:
Add explicit visibility for positive button in TagDialogFragment

danilo04 committed:
End files with a blank line

danilo04 committed:
Use binding as local variable in TagsActivity

For TagsActivity, we do not have to use a field to setup view binding.
We use a local variable instead and propagate the use of binding in
views and observers through making some functions extension functions of
binding.

danilo04 committed:
Rename test class to AddTagViewModelTest following ViewModel's name

danilo04 commented:
Howdy @ParaskP7 👋 . Thanks for the thorough review. I have added some changes answering your comments.

> The only problem I found during testing was during the Edit Tag testing, where a tag, that is not associated with any note, wouldn't update itself. Or two conflicting tags, that are not associated with any note, wouldn't merge as well. Can you double check that? 

This is a bug that already exists in the application. I am keeping it for now since there is a community PR that I think touches this bug. 

> Preparing the Java file, prior hitting the Convert to Kotlin button, might help you create a much better Kotlin file. For example, adding @Nullable or NonNull annotation wherever possible can end up helping you enormously. 

Good point! I will use this for future conversions.

> When converting to ViewBinding and removing all fields at the same time can end up being more complicated than expected. When doing that, do a commit-by-commit field conversion, with the first commit being the binding setup. 

Yes, I agree. Although I was using the tests during the refactoring, some refactors were merged together and that complicated the whole process (I think you could see that in the commit history). I need to be more careful for sure. 

> Last but not least, consider renaming the AddActivityViewModelTest test class to AddTagViewModelTest
Fixed in 4c72de558edac2616f48d7b66ceaa27d50e8a963.

ParaskP7 commented:
> Howdy @ParaskP7 👋 . Thanks for the thorough review. I have added some changes answering your comments.

👋 @danilo04 !

Thanks so much for applying all these changes! 🙏 

> This is a bug that already exists in the application. I am keeping it for now since there is a community PR that I think touches this bug.

Great, thanks for the clarification on that! 🙏 

> Good point! I will use this for future conversions.

🥇 🙏 

> Yes, I agree. Although I was using the tests during the refactoring, some refactors were merged together and that complicated the whole process (I think you could see that in the commit history). I need to be more careful for sure.

🥇 🙏 

> Fixed in 4c72de5.

👍

ParaskP7 reviewed (approved):
👋 @danilo04 !

I did another round on your latest changes and it LGTM, thanks so much for working on this PR, it is wild to see how much the tags screen is improved! ❤️ 

PS: Feel free to merge this PR when you are ready.
